runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -11963,3 +11986,295 @@ J9::Z::TreeEvaluator::tabortEvaluator(TR::Node * node, TR::CodeGenerator * cg)
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BL, node, doneLabel);
+   generateRXInstruction(cg, TR::InstOpCode::CLG, node, resultReg, topMemRef);
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BH, node, doneLabel);
+
+   if (!fieldUnresolved)
+      {
+      generateRXInstruction(cg, TR::InstOpCode::LAY, node, locationReg, generateS390MemoryReference(*loadMemRef, 0, cg));
+      }
+   // Save the object address to vmThread->floatTemp1 for jitReadBarrier helper.
+   generateRXInstruction(cg, TR::InstOpCode::STG, node, locationReg, generateS390MemoryReference(vmReg, offsetof(J9VMThread, floatTemp1), cg));
+
+   cg->generateDebugCounter(TR::DebugCounter::debugCounterName(cg->comp(), "readBar/helperCall"), 1, TR::DebugCounter::Cheap);
+   TR::LabelSymbol* callLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol* callEndLabel = generateLabelSymbol(cg);
+   TR::Instruction *gcPoint = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, callLabel);
+   gcPoint->setNeedsGCMap(0);
+   auto readBarHelperSnippet = new (cg->trHeapMemory()) TR::S390HelperCallSnippet(cg, node, callLabel,
+                                                                     cg->symRefTab()->findOrCreateRuntimeHelper(TR_readBarrier, false, false, false),
+                                                                     callEndLabel);
+   cg->addSnippet(readBarHelperSnippet);
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, callEndLabel);
+   // Reload the object after helper call.
+   generateRXInstruction(cg, loadOpCode, node, resultReg, generateS390MemoryReference(locationReg, 0, cg));
+   if (shouldShift)
+      {
+      generateRSInstruction(cg, TR::InstOpCode::SLLG, node, resultReg, resultReg, shiftAmount);
+      }
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, doneLabel, deps);
+   doneLabel->setEndInternalControlFlow();
+   cg->generateDebugCounter(TR::DebugCounter::debugCounterName(cg->comp(), "readBar/total"), 1, TR::DebugCounter::Cheap);
+
+   if (hasReturnReg)
+      {
+      node->setRegister(resultReg);
+      }
+   else
+      {
+      cg->stopUsingRegister(resultReg);
+      }
+
+   cg->stopUsingRegister(locationReg);
+   cg->stopUsingRegister(dummyRegForRA);
+   cg->stopUsingRegister(dummyRegForEntry);
+
+   return resultReg;
+#endif
+   }
+
+TR::Register *
+J9::Z::TreeEvaluator::arraycopyEvaluator(TR::Node * node, TR::CodeGenerator * cg)
+   {
+#ifdef OMR_GC_CONCURRENT_SCAVENGER
+   TR::Compilation* comp = cg->comp();
+
+   if (!TR::Compiler->om.shouldReplaceGuardedLoadWithSoftwareReadBarrier()  // No software concurrent scavenge
+           // For reference array copies that need an array store check always call the helper function
+           || !node->chkNoArrayStoreCheckArrayCopy()
+           || !node->isReferenceArrayCopy())
+      {
+      return OMR::Z::TreeEvaluator::arraycopyEvaluator(node, cg);
```

Comment:
I think we need to use a connector to call the base OMR API. @0xdaryl can you confirm if this is the call we really want or do we need something like:

```
return OMR::TreeEvaluatorConnector::arraycopyEvaluator(node, cg);
```